File: code/python/ch07/practice/practice_03_mod_exponentiation.py
"""
Practice 3: Modular Exponentiation
==============================
Chapter 7: Number Wizardry — Math for Programmers

PROBLEM
-------
Given three integers base, exp, and mod, compute (base^exp) % mod
using binary exponentiation (also known as fast exponentiation).

Do NOT use Python's built-in pow(base, exp, mod) — implement the
algorithm yourself to learn it!

Binary exponentiation works by squaring the base and halving the
exponent at each step. If the exponent is odd, multiply the result
by the current base.

INPUT FORMAT
------------
Three space-separated integers: base, exp, mod.

OUTPUT FORMAT
-------------
A single integer: (base^exp) % mod.

CONSTRAINTS
-----------
- 1 <= base <= 10^9
- 0 <= exp <= 10^18
- 2 <= mod <= 10^9

EXAMPLES
--------
Input:
  2 10 1000000007
Output: 1024

Input:
  2 20 1000000007
Output: 1048576

Input:
  123456789 0 1000000007
Output: 1

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(base: int, exp: int, mod: int) -> int:
    """Return base^exp mod m using binary exponentiation."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    base, exp, mod = map(int, input().split())
    print(solve(base, exp, mod))

File: code/cpp/ch07/practice/practice_03_mod_exponentiation.cpp
/*
 * Practice 3: Modular Exponentiation
 * ===================================
 * Chapter 7: Number Wizardry -- Math for Programmers
 *
 * PROBLEM:
 *   Compute (base^exp) % mod using binary exponentiation.
 *   This is MUCH faster than multiplying base*base*...*base exp times.
 *
 * EXAMPLES:
 *   solve(2, 10, 1000000007) -> 1024
 *   solve(2, 20, 1000000007) -> 1048576
 *   solve(123456789, 0, 1000000007) -> 1
 *   solve(2, 100, 1000000007) -> 976371285
 *
 * CONSTRAINTS:
 *   1 <= base <= 10^9
 *   0 <= exp <= 10^18
 *   2 <= mod <= 10^9
 *
 * INSTRUCTIONS:
 *   Replace "return 0;" with your solution.
 *   Use binary exponentiation: square base repeatedly, multiply
 *   into result when the current bit of exp is 1.
 *   Remember to take mod at EVERY multiplication to avoid overflow.
 */

#include <iostream>
using namespace std;

long long solve(long long base, long long exp, long long mod) {
    // TODO: Replace this with your solution
    return 0;
}

// -- Do not change anything below this line --------------------------
int main() {
    long long base, exp, mod;
    cin >> base >> exp >> mod;
    cout << solve(base, exp, mod) << endl;
    return 0;
}
